tools/ResxGen/ResxGen.psm1

```diff
@@ -0,0 +1,553 @@
+
+        /// <summary>
+        /// Gets the {1} describing the message resource for the specified event id.
+        /// </summary>
+        /// <param name="eventId">The event id for the message resource to retrieve.</param>
+        /// <returns>A {1} for the specified event id; otherwise, a null reference.</returns>
+        public static {1} GetMessage(int eventId)
+        {{
+            {1} result = null;
+            _resources.TryGetValue(eventId, out  result);
+            return result;
+        }}
+
+        const int MissingEventId = -1;
+
+        /// <summary>
+        /// Gets the {1} describing the a message resource used when an valid event id
+        /// is not found.
+        /// </summary>
+        public static {1} GetMissingEventMessage()
+        {{
+            return GetMessage(MissingEventId);
+        }}
+
+        static {1}()
+        {{
+            // Add the resource for mismatched event ids.
+            _resources.Add(MissingEventId, new {1}("MissingEventIdMessage", 1));
+
+'@
+
+# Adds an entry to the eventid -> resource name dictionary
+# String.Format arguments
+# {0} - class name
+# {1} - event id
+# {2} - the resource name for the event message
+# {3} - the number of parameters required to format the message. May be zero.
+[string] $codeEventEntryTemplate = @"
+
+            _resources.Add({1}, new {0}("{2}", {3}));
+"@
+
+# defines the end of the generated C# code.
+# String.Format arguments: None
+[string] $codeEpilogue = @"
+
+        }}
+
+        #endregion Static Fields and Methods
+    }}
+}}
+#endif
+"@
+
+#endregion C# code template strings
+
+<#
+  Provides a class for encapsulating a resource string entry from an ETW manifest
+#>
+class EventMessage
+{
+    #region properties
+
+    <#
+      Gets the message id.
+      This is used as a resource name.
+    #>
+    [string] $Id
+
+    <#
+      Gets the identifier used by an event to reference the message.
+    #>
+    [string] $EventReference
+
+    <#
+      The number of replaceable parameters in the message; from 0 through 99
+      Used to determine if string.Format is needed.
+    #>
+    [int] $ParameterCount
+
+    <#
+      Gets the message text
+    #>
+    [string] $Value
+
+    #endregion properties
+
+    <#
+    .SYNOPSIS
+      replaces FormatMessage format specifiers with String.Format equivalent.
+
+    .PARAMETER message
+      The message string to update.
+
+    .NOTES
+      See https://msdn.microsoft.com/en-us/library/windows/desktop/ms679351(v=vs.85).aspx.
+      Replaceable parameters are limited to %1 ... %99. Width and precision specifiers are
+      not currently supported since the manifest does not use them at the time of this writing.
+    #>
+    hidden [void] SetMessage([string] $message)
+    {
+        foreach ($source in [EventMessage]::escapeStrings.Keys)
+        {
+            if ($message.Contains($source))
```

Comment:
Calling `Contains` isn't really necessary - `Replace` won't create a new string if there's nothing to replace, and `Replace` has to search the string again, so it just does more work.

Reply:
And I think this whole loop would be easier to read/understand if it was something like this (basically inline the replacements since they aren't used elsewhere.)

```powershell
      $message = $message -replace '%t', '`t' `
                          -replace '%n', '`n' `
                          -replace '%r', '`r' `
                          -replace '%%', '`%' `
                          -replace '%space', ' ' `
                          -replace '%.', '.'
```

Reply:
Won't fix.